### PR TITLE
Support TypeSignature::Nullary

### PR DESCRIPTION
**Pull request by jayzhan211**
## Which issue does this PR close?

Closes #.

## Rationale for this change

Currently there are uniform(0), any(0), exact(0) 3 ways to describe zero argument signature. To avoid confusion, add `TypeSignature::ZeroArg` and disable 0 argument for those 3 signatures, they now require at least one argument.

### CODE REVIEW

File: datafusion/expr-common/src/signature.rs [hunk 1]
```diff
@@ -135,6 +134,8 @@ pub enum TypeSignature {
     /// Null is considerd as `Utf8` by default
     /// Dictionary with string value type is also handled.
     String(usize),
+    /// Zero argument
+    ZeroArg,
```
Comment (comphead) [inline datafusion/expr-common/src/signature.rs:138]: Maybe NoArgs would be more explanatory? 🤔

**Comment by findepi** [comment]
I think it would be nice to support a signature defined by something like `Vec<LogicalType>`
then ZeroArg / NoArgs would be just represented by an empty vec.
What do you think?

**Comment by jayzhan211** [comment]
> I think it would be nice to support a signature defined by something like `Vec<LogicalType>` then ZeroArg / NoArgs would be just represented by an empty vec. What do you think?

`Signature::coercible(vec![], Volatility::Immutable)` can represent this already. And you can also do Any(0), Uniform(0), so I want a standard representation for zero arg

File: datafusion/expr-common/src/signature.rs (hunk 1 above)
Comment (jayzhan211) [reply datafusion/expr-common/src/signature.rs:138]: I don't see the difference 👀

Comment (comphead) [reply datafusion/expr-common/src/signature.rs:138]: Its choice of wording yes. Another option we can take from https://en.wikipedia.org/wiki/Arity and name it Nullary to be in sync with unary, binary functions we already use.

**Review by comphead — APPROVED** [review]
lgtm thanks @jayzhan211 👍

**Review by Weijun-H — APPROVED** [review]
Ship it! Thanks @jayzhan211

**Comment by Weijun-H** [comment]
Thanks @2010YOUY01 and @comphead for review.

**Comment by alamb** [comment]
BTW I hit a usability rough edge due to this change I think: 
https://github.com/apache/datafusion/issues/13763

File: datafusion/expr-common/src/signature.rs [hunk 2]
```diff
@@ -135,6 +134,8 @@ pub enum TypeSignature {
     /// Null is considerd as `Utf8` by default
     /// Dictionary with string value type is also handled.
     String(usize),
+    /// Zero argument
+    NullAry,
```
Comment (kylebarron) [inline datafusion/expr-common/src/signature.rs:138]: Is the `A` here intended to be upper-case? I'd have expected `Nullary`

Comment (jayzhan211) [reply datafusion/expr-common/src/signature.rs:138]: `Nullary` looks good to me too

Comment (alamb) [reply datafusion/expr-common/src/signature.rs:138]: - I made a PR to fix: https://github.com/apache/datafusion/pull/13817

**Pull request merged**